### PR TITLE
bash: Additional common suggestions

### PR DESCRIPTION
**Pull request by habere-et-dispertire**
* Hash-pling
* Bash Strict Mode
* Static analysis

These suggestions are not exercise specific. Perhaps it would be helpful to have a root mentoring file to preface common guidance throughout the track.

### CODE REVIEW

File: tracks/bash/exercises/raindrops/mentoring.md [hunk 1]
```diff
@@ -25,6 +25,11 @@ echo "${output:-$1}"
 
 ## Common Suggestions
 
+* Check for a hash-pling: `#!/bin/bash`
```
Comment (IsaacG) [inline tracks/bash/exercises/raindrops/mentoring.md:28]: This is more commonly known as a shebang (hash bang).

File: tracks/bash/exercises/raindrops/mentoring.md [hunk 2]
```diff
@@ -25,6 +25,11 @@ echo "${output:-$1}"
...
+* Consider a more robust hash-pling: `#!/usr/bin/env bash` 
+* Aaron Maxwell's [Bash Strict Mode](http://redsymbol.net/articles/unofficial-bash-strict-mode/): `set -euo pipefail`
```
Comment (IsaacG) [inline tracks/bash/exercises/raindrops/mentoring.md:30]: The `set -e` is pretty controversial and has a load of pitfalls. Applying it blindly can be problematic.

**Review by IsaacG — CHANGES_REQUESTED** [review]

File: tracks/bash/exercises/raindrops/mentoring.md (hunk 1 above)
Comment (habere-et-dispertire) [reply tracks/bash/exercises/raindrops/mentoring.md:28]: # Bash Guidance


## Preflight checks

Check if the interpreter directive, aka hash-pling `#!` is present. Increased portability is acheived by not hard-coding this path. An indirect lookup with `env` is current best practice. Contrast:

```
#! /bin/bash
#! /usr/bin/env bash
```


## Interactive prompt self-help

Query if there an awareness of classic documentation that is typically available:

```
$ man  bash
$ info bash
```

A leaner alternative, [`tldr` (https://github.com/tldr-pages/tldr/)](https://github.com/tldr-pages/tldr/), reduces documentation down to common usage examples:

```
$ tldr bash

bash

Bourne-Again SHell, an `sh`-compatible command-line interpreter.
More information: <https://gnu.org/software/bash/>.

- Start an interactive shell session:
    bash

- Execute a command and then exit:
    bash -c "command"

...

```


## Static analysis & linting tools

[ShellCheck (https://shellcheck.net/)](https://shellcheck.net/) by Vidar Holen is a stable pre 1.0 tool that provides a set of clear, editable rationales for each type of detected issue. Further less problematic recommendations can also be requested:

```
$ shellcheck <file>
$ shellcheck --enable=all <file>
```


## Debug

Consider Aaron Maxwell's [Bash Strict Mode (http://redsymbol.net/articles/unofficial-bash-strict-mode/)](http://redsymbol.net/articles/unofficial-bash-strict-mode/) to help reduce an invisible issue down to a minimal working example:

```
set -Eeuo pipefail
IFS=$'\n\t'
```


## Resources, Books and Printables

A special mention must be made of Michael Mol's chrestomathy site, [Rosetta Code (https://rosettacode.org/wiki/Category:Bash)](https://rosettacode.org/wiki/Category:Bash). Comparing solutions with other known or even unknown languages can be rather instructive.

|                            | Address
---------------------------- | ---------------------
Manual                       | https://www.gnu.org/software/bash/manual/
Frequently Asked Questions   | http://mywiki.wooledge.org/BashFAQ](http://mywiki.wooledge.org/BashFAQ
Reference Card               | https://devhints.io/bash
Pure BASH Bible              | https://github.com/dylanaraps/pure-bash-bible](https://github.com/dylanaraps/pure-bash-bible
Essential Bash               | https://www.programming-books.io/essential/
BASH Notes for Professionals | https://www.programming-books.io/essential/bash/
PDF version, 1.8 MB          | https://books.goalkicker.com/BashBook/BashNotesForProfessionals.pdf

File: tracks/bash/exercises/raindrops/mentoring.md [hunk 3]
```diff
@@ -39,3 +39,80 @@ See the [two-fer notes](https://raw.githubusercontent.com/exercism/website-copy/
...
 
       [https://www.gnu.org/software/bash/manual/bash.html#Shell-Arithmetic](https://www.gnu.org/software/bash/manual/bash.html#Shell-Arithmetic)
+
+
+# Bash Guidance
+
+
+## Preflight checks
+
+Check if the interpreter directive, aka hash-pling `#!` is present. Increased portability is achieved by not hard-coding this path. An indirect lookup with `env` is current best practice. Contrast:
```
Comment (IsaacG) [inline tracks/bash/exercises/raindrops/mentoring.md:49]:
```suggestion
Check if the interpreter directive, aka hash-bash `#!` is present.
[Increased portability](http://mywiki.wooledge.org/BashGuide/CommandsAndArguments#Scripts) is achieved by not hard-coding this path, but looked up using `env`.
Contrast:
```

File: tracks/bash/exercises/raindrops/mentoring.md [hunk 4]
```diff
@@ -39,3 +39,80 @@ See the [two-fer notes](https://raw.githubusercontent.com/exercism/website-copy/
...
+## Interactive prompt self-help
+
+Query if there is an awareness of classic documentation that is typically available:
+
+```
+$ man  bash
+$ info bash
+```
+
+A leaner alternative, [`tldr` (https://github.com/tldr-pages/tldr/)](https://github.com/tldr-pages/tldr/), reduces documentation down to common usage examples:
```
Comment (IsaacG) [inline tracks/bash/exercises/raindrops/mentoring.md:66]: `tldr` isn't a standard tool. As such, it shouldn't be a "common" suggestion.

File: tracks/bash/exercises/raindrops/mentoring.md [hunk 5]
```diff
@@ -39,3 +39,80 @@ See the [two-fer notes](https://raw.githubusercontent.com/exercism/website-copy/
...
+    bash -c "command"
+
+...
+
+```
+
+
+## Static analysis & linting
+
+[ShellCheck (https://shellcheck.net/)](https://shellcheck.net/) by Vidar Holen is a stable pre 1.0 tool that provides a set of clear, editable rationales for each type of detected issue. Other less problematic recommendations can also be requested:
```
Comment (IsaacG) [inline tracks/bash/exercises/raindrops/mentoring.md:89]:
```suggestion
[ShellCheck (https://shellcheck.net/)](https://shellcheck.net/) by Vidar Holen is a tool that detects common issues
Other less problematic recommendations can also be requested:
```

File: tracks/bash/exercises/raindrops/mentoring.md [hunk 6]
```diff
@@ -39,3 +39,80 @@ See the [two-fer notes](https://raw.githubusercontent.com/exercism/website-copy/
...
+
+
+## Resources
+
+A special mention must be made of Michael Mol's chrestomathy site, [Rosetta Code (https://rosettacode.org/wiki/Category:Bash)](https://rosettacode.org/wiki/Category:Bash). Comparing solutions with other known or even unknown languages can be rather instructive.
+
+|                                  | Address
+---------------------------------- | ---------------------
+Manual                             | https://www.gnu.org/software/bash/manual/
+Frequently Asked Questions         | http://mywiki.wooledge.org/BashFAQ
```
Comment (IsaacG) [inline tracks/bash/exercises/raindrops/mentoring.md:114]:
```suggestion
Wooledge Wiki         | http://mywiki.wooledge.org/
```

File: tracks/bash/exercises/raindrops/mentoring.md [hunk 7]
```diff
@@ -39,3 +39,80 @@ See the [two-fer notes](https://raw.githubusercontent.com/exercism/website-copy/
...
+```
+set -Eeuo pipefail
+IFS=$'\n\t'
+```
+
+
+## Resources
+
+A special mention must be made of Michael Mol's chrestomathy site, [Rosetta Code (https://rosettacode.org/wiki/Category:Bash)](https://rosettacode.org/wiki/Category:Bash). Comparing solutions with other known or even unknown languages can be rather instructive.
+
```
Comment (IsaacG) [inline tracks/bash/exercises/raindrops/mentoring.md:110]: https://wiki.bash-hackers.org/scripting/tutoriallist might be worth a mention, as a curated/ranked list of tutorials.

**Review by IsaacG — CHANGES_REQUESTED** [review]

File: tracks/bash/exercises/raindrops/mentoring.md (hunk 2 above)
Comment (habere-et-dispertire) [reply tracks/bash/exercises/raindrops/mentoring.md]: I removed it and mentioned it under Debug as a possible way to get a minimal working example.

File: tracks/bash/exercises/raindrops/mentoring.md (hunk 7 above)
Comment (habere-et-dispertire) [reply tracks/bash/exercises/raindrops/mentoring.md:113]: I'm adding it.

**Comment by habere-et-dispertire** [comment]
I see that this repository is no longer seeking [unsolicited contributions](https://github.com/exercism/website/blob/150d2aba499b1f9e4f358297c201ba91e629fdb0/docs/CONTRIBUTING.md).

Should I go ahead and close this pull request?

**Comment by IsaacG** [comment]
These suggestions are all generic to bash in general and not specific to this exercise.

The mentoring notes are supposed to provide exercise specific guidance to mentors; I don't think it makes sense to include general guidance to every exercises' mentoring file. There might be a better place for that, but I'm not sure where that would be.

Additionally, these notes aren't actually related to providing guidance on this exercise. Or providing guidance at all. They seem more like general notes for someone writing bash, not mentoring bash, which again makes them feel like there's probably a better place for them.

@glennj @kotp  Any thoughts?

**Comment by glennj** [comment]
I agree that these particular changes are not appropriate for the mentor notes for this exercise. Some of the suggestions are not without controversy (i.e. I have opinions about "strict mode")

I've never been a fan of the mentor notes. I think notes for each exercise build on all the previous exercises. There are certainly particular "classes" of errors that students make that are indeed common. I keep my personal notes in git: https://github.com/glennj/exercism.io/tree/main/_mentor_notes

IMO a wiki would be a more ideal format for track notes.

**Comment by habere-et-dispertire** [comment]
> I agree that these particular changes are not appropriate for the mentor notes for this exercise.

My original intention was for it to act as a resource/guide for first timers/beginners as well as mentors. The notes are not intended to be exercise specific and I couldn't see where to put them. I was just exploring Exercism and found the mentor notes tab. (I'm still learning where things go in this large ecosystem.) My original comment at top:

"These suggestions are not exercise specific. Perhaps it would be helpful to have a root mentoring file to preface common guidance throughout the track."

**Comment by IsaacG** [comment]
The mentoring notes are specifically intended for mentors to provide guidance on mentoring an exercise. 

I don't think there's currently a place for general development/debugging guidance. I'm also not sure this is the ideal platform for that sort of guide. If we go down the route of providing that sort of guide, how would that look for other languages?

**Comment by kotp** [comment]
If the "Hash-Pling" is `#!`, then for most of these exercises that is not applicable.  Are they not "libraries" instead, meant to be loaded in the environment, rather than executed?

We should indeed mentor that this is not needed and the reasons why, when we find students doing this kind of thing.  I think that the same is for any of the things that Students are doing, regardless of "style guides".  The way that styles improve is through thought, rather than blind following of such things.

Otherwise, we often cut the ends off of roasts just because that is how Grandma always did it, which is not a great reason to continue doing things.  We should understand that the reason Grandma did that was because otherwise it would simply not fit in the pan.

**Comment by glennj** [comment]
I think until Exercism leadership decides what to do about a support forum, this kind of knowledge doesn't really have a place on the website.

Perhaps you could find a place to add it [in the Bash track documentation](https://exercism.org/docs/tracks/bash)

I think this PR can be closed.

**Comment by kotp** [comment]
I thought the support forum was the mentoring requests per exercise, especially for this kind of advice.  And support for how to get there is on Gitter.

I think it can be closed (and is at this point) as well.  I do have a question though, @glennj , if we put it in the Bash track documentation, does it not get placed on the website, so is this contradictory advice?  I know it is easily done, but I am a bit confused at the moment.

**Comment by glennj** [comment]
I suppose I was (indirectly) suggesting that habere-et-dispertire could review the existing track documentation to see if (some of) the suggested content to be placed there.

**Pull request closed without merging**